### PR TITLE
Accelerate SDPA on Arm CPUs: Implement fast exp in AdvSIMD vectorizer

### PR DESCRIPTION
**Pull request by fadara01**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #177012
* #177009
* __->__ #176881

Similar to #151441 - this PR adds an ASIMD implementation for fast exponential intended for cases where outputs will be downcasted to FP16 / BF16 (e.g. attention softmax).

Implementation is similar to the existing exp_u20, but:
  - uses a third degree polynomial approximation for exp(r) instead of a
    fifth degree one, with coefficients re-tuned.
  - does not split natural log (ln) into high / low parts
  - clamps exp(x) to 0 for x < -87.346351f and inf for x > 88.3762589f

Overall, this allows us to ditch 4 FMLAs (out of 7 in the current impl) and the nasty if branch

## Accuracy

Following a similar approach to #151441, I tested accuracy for fast exp using [this](https://gist.github.com/fadara01/1f0a7bc4b63193102b5f6b0283a9b02f) script. The script iterates over all possible FP32 bit patterns and calculates ULP between:
- `fexp_u20` with inputs in FP32, outputs converted to BF16/FP16
- `std::exp` with inputs in FP32, outputs converted to BF16/FP16  

Accuracy script shows that fast exp is:
- for FP16: accurate within a maximum of 1 FP16 ULPs
- for BF16: accurate within a maximum of 1 BF16 ULPs for inputs in [-87.346351, 88.376] - we clamp inputs outside this range to 0 / inf.

## Performance

Using [this SDPA benchmark](https://gist.github.com/fadara01/5357a52299a3722587f6691d145e71e9), here are the scaled-dot-production-attention speedups achieved with 16 Neoverse-V2 cores:

| B | Hq | Hkv | Lq | Lk | D | causal | gqa | Speedup vs current |
|---:|---:|---:|---:|---:|---:|---|---|---:|
| 1 | 32 | 8 | 2048 | 2048 | 128 | True | True | +9.48% |
| 1 | 32 | 8 | 1 | 2048 | 128 | False | True | -1.42% (noise) |
| 1 | 16 | 16 | 6400 | 6400 | 80 | False | False | +5.18% |
| 1 | 20 | 20 | 1500 | 1500 | 64 | False | False | +6.63% |
| 8 | 20 | 20 | 1500 | 1500 | 64 | False | False | +9.31% |

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @drisspg @liangel-02 @howardzhang-cv

### CODE REVIEW

File: aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h [hunk 1]
```diff
@@ -354,7 +354,43 @@ class Vectorized<float> {
...
+    // exp_u20, but:
+    // - uses a third degree polynomial approximation for exp(r) instead of a
+    // fifth degree one, with coefficients re-tuned.
+    // - does not split natural log (ln) into high / low parts
+    // - clamps exp(x) to 0 for x < -87.6831131f and inf for x > 88.3762589f
+
+    const float32x4_t lower_bound = vdupq_n_f32(-0x1.5ebb82p+6f);
+    const float32x4_t upper_bound = vdupq_n_f32(0x1.61814ap+6f);
+    const float32x4_t inv_ln2 = vdupq_n_f32(0x1.715476p+0f);
+    const float ln2 = 0x1.62e43p-1f;
```
Comment (Skylion007) [inline aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h]: Nit: should be constexpr

File: aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h [hunk 2]
```diff
@@ -354,7 +354,43 @@ class Vectorized<float> {
...
+    const float c2 = 0x1.5592ecp-3f;
```
Comment (Skylion007) [inline aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h]: Should also be constexpr

File: aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h (hunk 1 above)
Comment (fadara01) [reply aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h]: well spotted, done

File: aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h (hunk 2 above)
Comment (fadara01) [reply aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h]: done

File: aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h [hunk 3]
```diff
@@ -354,7 +354,47 @@ class Vectorized<float> {
...
+    const float32x4_t c3 = vdupq_n_f32(0x1.017d34p-1f);
+    const uint32x4_t lt_lower = vcltq_f32(values, lower_bound);
+    const uint32x4_t gt_upper = vcgtq_f32(values, upper_bound);
+
+    // exp(x) = 2^n (1 + exp(r))
+    // r = x - n*ln2, with n = round(x/ln2)
+    // exp(r) ~ poly(r) = r + r^2 * (c3 + c2 * r)
+
+    // n = round(x / ln2), r = x - n*ln2
+    float32x4_t n = vrndaq_f32(vmulq_f32(values, inv_ln2));
```
Comment (jgong5) [inline aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h:381]: Shall we use `vrndnq_f32` which is RNE here?

Comment (fadara01) [reply aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h:381]: Thanks for the review @jgong5 :)
I used vrndaq_f32 intentionally here for consistency with the [existing NEON exp_u20 implementation](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h#L338) in PyTorch, which also uses vrndaq_f32. It also matches the AArch64 AdvSIMD expf implementation in glibc.
I kept the same tie-breaking behavior in this kernel, and the ULP measurements I collected were with that behavior.

Do you have strong opinions on this?

**Review by Skylion007 — APPROVED** [review]

File: aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h (hunk 3 above)
Comment (jgong5) [reply aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h:381]: > Do you have strong opinions on this?

No, keeping the original behavior sounds good to me.

**Review by jgong5 — APPROVED** [review]
